recipes/dcase2021_task4_separation_SED/train_sed.py

```diff
@@ -0,0 +1,168 @@
+
+
+def single_run(config, log_dir, gpus, checkpoint_resume=""):
+
+    config.update({"log_dir": log_dir})
+
+    ##### data prep ##########
+    encoder = ManyHotEncoder(
+        list(classes_labels.keys()),
+        audio_len=config["data"]["audio_max_len"],
+        frame_len=config["feats"]["n_filters"],
+        frame_hop=config["feats"]["stride"],
+        net_pooling=config["data"]["net_subsample"],
+        fs=config["data"]["fs"],
+    )
+
+    synth_set = StronglyAnnotatedSet(
+        config["data"]["synth_folder"],
+        config["data"]["synth_tsv"],
+        encoder,
+        train=True,
+        target_len=config["data"]["audio_max_len"],
+    )
+
+    weak_set = WeakSet(
+        config["data"]["weak_folder"],
+        config["data"]["weak_tsv"],
+        encoder,
+        target_len=config["data"]["audio_max_len"],
+    )
+
+    unlabeled_set = UnlabelledSet(
+        config["data"]["unlabeled_folder"],
+        encoder,
+        target_len=config["data"]["audio_max_len"],
+    )
+
+    synth_val = StronglyAnnotatedSet(
+        config["data"]["synth_val_folder"],
+        config["data"]["synth_val_tsv"],
+        encoder,
+        train=False,
+        return_filename=True,
+        target_len=config["data"]["audio_max_len"],
+    )
+
+    weak_eval = WeakSet(
+        config["data"]["weak_val_folder"],
+        config["data"]["weak_val_tsv"],
+        encoder,
+        target_len=config["data"]["audio_max_len"],
+        return_filename=True,
+        train=False,
+    )
+
+    public_eval = StronglyAnnotatedSet(
+        config["data"]["pub_eval_folder"],
+        config["data"]["pub_eval_tsv"],
+        encoder,
+        train=False,
+        return_filename=True,
+        target_len=config["data"]["audio_max_len"],
+    )
+
+    tot_train_data = [synth_set, weak_set, unlabeled_set]
+    train_dataset = torch.utils.data.ConcatDataset(tot_train_data)
+
+    batch_sizes = config["training"]["batch_size"]
+    samplers = [torch.utils.data.RandomSampler(x) for x in tot_train_data]
+    batch_sampler = ConcatDatasetBatchSampler(samplers, batch_sizes)
+
+    valid_dataset = torch.utils.data.ConcatDataset([weak_eval, synth_val, public_eval])
+
+    epoch_len = min(
+        [
+            len(tot_train_data[indx])
+            // (
+                config["training"]["batch_size"][indx]
+                * config["training"]["accumulate_batches"]
+            )
+            for indx in range(len(tot_train_data))
+        ]
+    )
+
+    ##### models and optimizers  ############
+
+    sed_student = CRNN(**config["net"])
+    sed_teacher = deepcopy(sed_student)
+
+    opt = torch.optim.Adam(sed_student.parameters(), 1e-8, betas=(0.9, 0.999))
+    exp_steps = config["training"]["n_epochs_warmup"] * epoch_len
+    exp_scheduler = {
+        "scheduler": ExponentialWarmup(opt, config["opt"]["lr"], exp_steps),
+        "interval": "step",
+    }
+
+    scaler = TorchScaler("instance", "minmax", (1, 2))
```

Comment:
General thinking: 
For first experiment, do we need to put the scaler the same way as before (using the all dataset ? ) since I already make the experiment of how much it changes the performances in last year's code 🤔

Reply:
Right now performance is decent even with min-max. For separation also it is useful to have this one. 
Maybe if we want to compare directly we need however the dataset scaler. Is the priority to make a comparison (which we can do later maybe) or deliver a baseline with last year plateau performance ?

Reply:
The priority is to give the baseline, but it is important we have a list of everything that is changed from last year.
This scaler change, is mostly "useful" as you say so it is easy to use SSep with it, it seems we have an idea of how much it changes, so let's do the experiment later, but let's keep in mind we'll probably want to do it for the paper explaining the task + baseline. :)